File: frontend/src/app/services/tank.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Tank {
  id?: number;
  farmId: number;
  name: string;
  capacity?: number;
  material?: string;
  height?: number;
  status: string;
  geometry: any; // GeoJSON Point
  photoUrl?: string;
}

@Injectable({
  providedIn: 'root'
})
export class TankService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = '/api/tanks';

  getAllTanks(): Observable<Tank[]> {
    return this.http.get<Tank[]>(this.apiUrl);
  }

  getTanksByFarm(farmId: number): Observable<Tank[]> {
    return this.http.get<Tank[]>(`${this.apiUrl}/farm/${farmId}`);
  }

  createTank(tank: Tank): Observable<Tank> {
    return this.http.post<Tank>(this.apiUrl, tank);
  }

  updateTank(id: number, tank: Tank): Observable<Tank> {
    return this.http.put<Tank>(`${this.apiUrl}/${id}`, tank);
  }

  deleteTank(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: backend/src/main/java/com/company/farmdigitaltwin/service/TankService.java
package com.company.farmdigitaltwin.service;

import com.company.farmdigitaltwin.dto.TankRequest;
import com.company.farmdigitaltwin.entity.Farm;
import com.company.farmdigitaltwin.entity.Tank;
import com.company.farmdigitaltwin.exception.ResourceNotFoundException;
import com.company.farmdigitaltwin.repository.FarmRepository;
import com.company.farmdigitaltwin.repository.TankRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class TankService {

    private final TankRepository tankRepository;
    private final FarmRepository farmRepository;

    public TankService(TankRepository tankRepository, FarmRepository farmRepository) {
        this.tankRepository = tankRepository;
        this.farmRepository = farmRepository;
    }

    @Transactional(readOnly = true)
    public List<Tank> getAllTanks() {
        return tankRepository.findAll();
    }

    @Transactional(readOnly = true)
    public List<Tank> getTanksByFarm(Long farmId) {
        return tankRepository.findByFarmId(farmId);
    }

    @Transactional(readOnly = true)
    public Tank getTankById(Long id) {
        return tankRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Tank", "id", id));
    }

    @Transactional
    public Tank createTank(TankRequest request) {
        Farm farm = farmRepository.findById(request.getFarmId())
                .orElseThrow(() -> new ResourceNotFoundException("Farm", "id", request.getFarmId()));

        Tank tank = Tank.builder()
                .farm(farm)
                .name(request.getName())
                .capacity(request.getCapacity())
                .material(request.getMaterial())
                .height(request.getHeight())
                .status(request.getStatus() != null ? request.getStatus() : "OPERATIONAL")
                .geometry(request.getGeometry())
                .photoUrl(request.getPhotoUrl())
                .build();

        return tankRepository.save(tank);
    }

    @Transactional
    public Tank updateTank(Long id, TankRequest request) {
        Tank tank = getTankById(id);
        Farm farm = farmRepository.findById(request.getFarmId())
                .orElseThrow(() -> new ResourceNotFoundException("Farm", "id", request.getFarmId()));

        tank.setFarm(farm);
        tank.setName(request.getName());
        tank.setCapacity(request.getCapacity());
        tank.setMaterial(request.getMaterial());
        tank.setHeight(request.getHeight());
        if (request.getStatus() != null) tank.setStatus(request.getStatus());
        tank.setGeometry(request.getGeometry());
        tank.setPhotoUrl(request.getPhotoUrl());

        return tankRepository.save(tank);
    }

    @Transactional
    public void deleteTank(Long id) {
        Tank tank = getTankById(id);
        tankRepository.delete(tank);
    }
}
